viz/LaserLine.cpp

```diff
@@ -2,44 +2,154 @@
+    return lineHolder->getRootNode();
 }
 
-void LaserLine::updateMainNode ( osg::Node* node )
+void LaserLine::updateMainNode (osg::Node* node)
 {
-    osg::Geode* geode = static_cast<osg::Geode*>(node);
-    // Update the main node using the data in p->data
+    lineHolder->update(vertices, color);
 }
 
 void LaserLine::updateDataIntern(base::samples::LaserScan const& value)
 {
-    p->data = value;
-    std::cout << "got new sample data" << std::endl;
+    loadLineVerticesFromLaserScan(value);
 }
 
+bool LaserLine::isShowRange() const
+{
+    return lineHolder->isShowRange();
+}
+void LaserLine::setShowRange(bool value)
+{
+    lineHolder->setShowRange(value);
+}
+
+QColor LaserLine::getColor()
+{
+    QColor qcolor;
+    qcolor.setRgbF(color[0], color[1], color[2], color[3]);
+    return qcolor;
+}
+
+void LaserLine::setColor(QColor value)
+{
+    if (value.isValid()) {
+        color.set(value.redF(),
+                  value.greenF(),
+                  value.blueF(),
+                  value.alphaF());
+    }
+}
+
+double LaserLine::getLineWidth()
+{
+    return (double)lineHolder->getLineWidth();
+}
+
+void LaserLine::setLineWidth(double lineWidth)
+{
+    lineHolder->setLineWidth((float)lineWidth);
+}
+
+bool LaserLine::getShowAllLines()
+{
+    return showAllLines;
+}
+
+void LaserLine::setShowAllLines(bool value)
+{
+    showAllLines = value;
+}
+
+osg::ref_ptr<osg::Vec3Array> LaserLine::computePoints(base::samples::LaserScan const& laserScan, std::vector<bool>& validity)
+{
+    osg::Vec3Array *points = new osg::Vec3Array();
+
+    std::vector<uint32_t> ranges = laserScan.ranges;
+
+    double angle = laserScan.start_angle; //start angle
+    double angleStep = laserScan.angular_resolution; //angle step
+
+    for (int i = 0; i < ranges.size(); i++) {
+
+        //compute the angle rotation
+        Eigen::Quaterniond rot = Eigen::Quaterniond(Eigen::AngleAxisd(angle + (i * angleStep), Eigen::Vector3d::UnitZ()));
+        Eigen::Vector3d pt =  rot * Eigen::Vector3d(milimeters_to_meters(ranges[i]), 0.0, 0.0);
+
+        //laser line point
+        points->push_back(osg::Vec3f(pt[0], pt[1], pt[2]));
+
+        //check if range is valid
+        validity.push_back(laserScan.isRangeValid(ranges[i]));
+    }
+
+    return points;
+}
+
+void LaserLine::computeLines(osg::Vec3Array& lineVertices, osg::Vec3Array const& points, std::vector<bool> const& validity) {
+
+    const double maxDt = 2.5;
+
+    lineVertices.clear();
+
+    for (int i = 0; i < points.size()-1; i++) {
+
+        if (!validity[i+0] || !validity[i+1]){
+            continue;
+        }
+
+        //compute Euclidean distance between two points
+        double dtx =  fabs(points[i+0][0] - points[i+1][0]);
+        double dty =  fabs(points[i+0][1] - points[i+1][1]);
+        double dtz =  fabs(points[i+0][2] - points[i+1][2]);
+        double dt = sqrt(dtx * dtx + dty * dty + dtz * dtz);
+
+        //if the distance is bigger than max distance, then this line is dropped
```

Comment:
Finally ... I understand what you are trying to do ... You are filtering points that look like they have nothing "solid"  in-between.

*All* linear algebra implementations have ways to compute norms. According to [the OSG API documentation](http://trac.openscenegraph.org/documentation/OpenSceneGraphReferenceDocs/a00974.html),

~~~
double dt = (points[i+0] - points[i+1]).length();
~~~